### PR TITLE
packaging: bring down the delta between 14.04 and 16.04.

### PR DESCRIPTION
**Pull request by chipaca**
This is rather aggressive, and needs further testing before we're sure
it's the way to go, but one can hope.

### CODE REVIEW

**Comment by chipaca** [comment]
Hmm, perhaps _too_ aggressive :-)
iterating.

File: packaging/ubuntu-14.04/rules [hunk 1]
```diff
@@ -141,7 +141,7 @@ override_dh_auto_test:
 # a tested default (production) build should have no test keys
 ifeq (,$(filter nocheck,$(DEB_BUILD_OPTIONS)))
 	# check that only the main trusted account-keys are included
-	[ $$(strings _build/bin/snapd|grep -c -E "public-key-sha3-384: [a-zA-Z0-9_-]{64}") -eq 1 ]
+	[ $$(strings _build/bin/snapd|grep -c -E "public-key-sha3-384: [a-zA-Z0-9_-]{64}") -eq 2 ]
```
Comment (zyga) [inline packaging/ubuntu-14.04/rules]: Why are there two keys there now?

File: packaging/ubuntu-14.04/rules [hunk 2]
```diff
@@ -174,6 +172,7 @@ override_dh_install:
 	# it's a conf file so we're stuck with it
 	mv debian/snapd/etc/profile.d/snapd.sh debian/snapd/etc/profile.d/apps-bin-path.sh
 
+	# trusty needs this to make /snap rshared
 	install --mode=0644 debian/snap.mount.service debian/snapd/$(SYSTEMD_UNITS_DESTDIR)
```
Comment (zyga) [inline packaging/ubuntu-14.04/rules]: Hmmm, hold on, didn't we do that from snap-confine and de-comissioned this unit file?

File: packaging/ubuntu-14.04/gbp.conf [hunk 3]
```diff
@@ -1,4 +0,0 @@
-[DEFAULT]
```
Comment (zyga) [inline packaging/ubuntu-14.04/gbp.conf]: Thanks for removing a dead file.

File: packaging/ubuntu-14.04/snapd.postrm [hunk 4]
```diff
@@ -4,27 +4,24 @@ set -e
 
 systemctl_stop() {
     unit="$1"
-    if systemctl is-active -q "$unit"; then
-        echo "Stoping $unit"
+    for i in $(seq 10); do
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: I think this warrants a comment.

File: packaging/ubuntu-14.04/snapd.postrm [hunk 5]
```diff
@@ -4,27 +4,24 @@ set -e
...
-            continue
-        fi
-        echo "Stopping $unit"
-        systemctl_stop "$unit"
-    done
+    # snap.mount.service is a trusty thing
+    systemctl_stop snap.mount.service
 
-    systemctl stop snap.mount.service || true
+    units=$(systemctl list-unit-files --full | grep '^snap[-.]' | cut -f1 -d ' ' | grep -vF snap.mount.service || true)
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: Note that this doesn't catch units like `var-lib-snapd-snap-$SNAP_NAME`

Comment (chipaca) [reply packaging/ubuntu-14.04/snapd.postrm]: what uses those?

File: packaging/ubuntu-14.04/rules (hunk 2 above)
Comment (chipaca) [reply packaging/ubuntu-14.04/rules]: no idea. Did we?

File: packaging/ubuntu-14.04/rules (hunk 1 above)
Comment (chipaca) [reply packaging/ubuntu-14.04/rules]: I think one is the old canonical one, and the other is the new generic one pedronis' been working on.

File: packaging/ubuntu-14.04/rules (hunk 2 above)
Comment (mvo5) [reply packaging/ubuntu-14.04/rules]: We do this from snap-confine now, but I'm not sure we can de-commission the unit file just yet. We still have this problem that snap-confine runs for the first time *after* all the snaps are mounted (we still have this problem on 16.04 and its causing trouble with lxd).

File: packaging/ubuntu-16.04/rules [hunk 6]
```diff
@@ -146,68 +146,6 @@ ifeq (,$(filter nocheck,$(DEB_BUILD_OPTIONS)))
 	$(MAKE) -C cmd check
 endif
 
-override_dh_systemd_enable:
```
Comment (mvo5) [inline packaging/ubuntu-16.04/rules]: Woah! :heart:

**Comment by mvo5** [comment]
@zyga What particular issue did you see with dpkg when symlinks are used? AFAICT this works fine. As for generating from a template: some of the things we use (autopkgtest) are very inflexible. They expect a git tree with debian/ packaging. I.e. autopkgtest will checkout the tree, run dpkg-buildpackage and after that run the tests. Last I checked there was no way to run any pre-generation script. Except for maybe hacking debian/rules clean in interessting ways to generate the real rules file and proxy all calls to them. Not sure if we want to go down that route (even though it would be interessting).

File: packaging/ubuntu-14.04/snapd.postrm (hunk 5 above)
Comment (zyga) [reply packaging/ubuntu-14.04/snapd.postrm]: Those get created by systemd if snap mount dir is different. I see that this is just for specific version of Ubuntu but I wanted to highlight this.

**Pull request merged**